### ISSUE TITLE
Intake for image files

### ISSUE THREAD

**Issue by quartox** [intake/intake#342]
I am starting a proof of concept for how to use intake to make a catalog of images for deep learning. My users have told me they think it makes the most sense to have the catalog return the filenames so that they can use tensorflow or pytorch or [insert framework here] to load the images. I talked to stanley at pycon and he suggested I open an issue to discuss this usecase in more detail.

**Comment by martindurant** [comment]
Certainly!
I am about to go rural for the weekend, so I will give you a very brief opinion.

Images are certainly array-like things - that's what the data fundamentally is - and it would make sense to view one image or a group of images as a dataset. Generally, Intake cares about turning dataset prescriptions into in-memory (or possibly dask-lazy) containers of the same data. In this sense, it would not be doing what you suggest, we would care about how to actually turn each image file into data.

However, some of the drivers have extra methods on them, such as `to_spark`, which bypasses whatever loading mechanism there may be, and hands off the details to some other framework. Thus, I imagine you could well have a dataset which is a set of images, `.discover()` tells you their metadata (number, size...), `to_dask()` produces a lazy dask array, `read()` produces a numpy array, and `.to_tf()` produces whatever you need to hand to TensorFlow; or even the TF object created from that (without loading). Incidentally, this might give the the opportunity to benefit from caching, downloading the files from a remote store (and maybe decompressing) upon first access.

Alternatively, it could be possible to view the list of filenames/glob as a dataset of its own, in which case `read()` would return only that, and you can do whatever you like with it. This would be the very simplest of drivers, but not entirely fit in with the style of others.

**Comment by quartox** [comment]
Could add a `.to_filepaths()` method to a driver to only return the file names and still use the built in caching?

I can imagine a scenario where I can convince people to add a method to our driver for each framework, but it sounds like they want this as a debugging tool at least.

The other aspect I am interested in is data versioning, which may be best discussed in another issue. At first I was thinking about logging all of the file names in git with an md5 sum of the file itself to ensure reproducibility. But is there another mechanism for intake to help reproduce results? I am specifically thinking of a case where the underlying number of images has grown but I want to read in the same ones I used 6 months ago.

**Comment by martindurant** [comment]
> Could add a .to_filepaths() method to a driver

Certainly. The implementation can choose to reference cached files indeed.

> I am interested in is data versioning

There is some discussion on this already ( #351 ), but no concrete plans yet. A number of related aspects should be considered:
- persisted data, keyed by timestamp when last loaded from source
- cached files, with some checksum from remote (but only if the remote exposes these checksums! You don't want to download whole files just to check if they have been updated) or simple timestamp. Storing file lists only is easier, but still needs structuring.
- external version control, such as DAT and git. The former is already possible, and I believe @mmccarty  plans to implement the latter as a POC. The caching mechanism should expose the labels of the versions and the user should be able to pick one, or "latest".

In the first two cases, Intake would be in control of keeping/expunging previous versions. That will be more than you can do with simple YAML files, maybe would need an internal database.

**Comment by satrya-sabeni** [comment]
Have you found a good way to save image datasets with Intake?

**Comment by martindurant** [comment]
Intake is not normally in the business of saving data, but of loading data. For "persist" and "export" features, any given data type is converted to the single canonical format for the type. For images, which are nd-arrays, this type will be zarr. That's a very convenient format for a number of reasons, but not one that any image-specific framework will understand.
